clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -4170,15 +4244,17 @@ void Sema::ConstructOpenCLKernel(FunctionDecl *KernelCallerFunc,
   // Visit handlers to generate information for optimization record only if
   // optimization record is saved.
   if (!getLangOpts().OptRecordFile.empty()) {
-    Visitor.VisitRecordBases(KernelObj, kernel_decl, kernel_body, int_header,
-                             int_footer, opt_report);
-    Visitor.VisitRecordFields(KernelObj, kernel_decl, kernel_body, int_header,
-                              int_footer, opt_report);
+    Visitor.VisitRecordBases(KernelObj, argsSizeChecker, kernel_decl,
+                             kernel_body, int_header, int_footer, opt_report,
+                             esimdKernel);
+    Visitor.VisitRecordFields(KernelObj, argsSizeChecker, kernel_decl,
+                              kernel_body, int_header, int_footer, opt_report,
+                              esimdKernel);
```

Comment:
esimdKernel should probably be called right after/before argsSizeChecker since its a diagnostic checker. This should happen before code generation visitors